tools/common/Assembly.cs

```diff
@@ -417,4 +430,95 @@ public IEnumerable<string> GetRelatedFiles ()
 			}
 		}
 	}
+
+	public class AssemblyCollection : IEnumerable<Assembly>
+	{
+		Dictionary<string, Assembly> HashedAssemblies = new Dictionary<string, Assembly> (StringComparer.OrdinalIgnoreCase);
+
+		public void Add (Assembly assembly)
+		{
+			Assembly other;
+			if (HashedAssemblies.TryGetValue (assembly.Identity, out other))
+				throw ErrorHelper.CreateError (2017, "The assembly '{0}' is referenced from two different locations: '{1}' and '{2}'.", assembly.Identity, other.FullPath, assembly.FullPath);
+			HashedAssemblies.Add (assembly.Identity, assembly);
```

Comment:
Bug #51670 is such a case (but different issue) where I was thinking we should be _warning_ when this happens (to make it clear our SDK has precedence).